ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/transaction/arjunacore/jca/TransactionImporterImple.java

```diff
@@ -211,17 +196,85 @@ public void removeImportedTransaction(Xid xid) throws XAException
 	}
 	
 	public Set<Xid> getInflightXids(String parentNodeName) {
-		Iterator<TransactionImple> iterator = _transactions.values().iterator();
+		Iterator<AtomicReference<TransactionImple>> iterator = _transactions.values().iterator();
 		Set<Xid> toReturn = new HashSet<Xid>();
 		while (iterator.hasNext()) {
-			TransactionImple next = iterator.next();
-			if (next.getParentNodeName().equals(parentNodeName)) {
-				toReturn.add(next.baseXid());
+			TransactionImple imported = extractImportedTransaction(iterator.next());
+
+			if (imported != null && imported.getParentNodeName().equals(parentNodeName)) {
+				toReturn.add(imported.baseXid());
 			}
 		}
 		return toReturn;
 	}
 
-	private static ConcurrentHashMap<SubordinateXidImple, TransactionImple> _transactions = new ConcurrentHashMap<SubordinateXidImple, TransactionImple>();
+	private TransactionImple addImportedTransaction(TransactionImple importedTransaction, Xid mapKey, Xid xid, int timeout) {
+		SubordinateXidImple importedXid = new SubordinateXidImple(mapKey);
+		// We need to store the imported transaction in a volatile field holder so that it can be shared between threads
+		AtomicReference<TransactionImple> holder = new AtomicReference<>();
+		AtomicReference<TransactionImple> existing;
+
+		if ((existing = _transactions.putIfAbsent(importedXid, holder)) != null) {
+			holder = existing;
+		}
+
+		TransactionImple txn = holder.get();
+
+		if (txn == null) {
+			// retry the get under a lock - this double check idiom is safe because AtomicReference is effectively
+			// a volatile so can be concurrently accessed by multiple threads
+			synchronized (holder) {
+				txn = holder.get();
+				if (txn == null) {
```

Comment:
OK, looks fine to me then. I thought you were saying previously that the holder atomicreference could point to different objects at any point (which it can, but I read it as though it applied in this specific code) so I was thinking that synchronizing on it would be pointless. If I follow you correctly you are saying by the time the synchronized (holder) is called the object that the atomicreference points at is fixed. The wording in the comment of "effectively a volatile" actually applies in a beneficial sense - an initial reading of it makes the code look scarier but actually it means that synchronizing on the holder (declared as a local variable) will sync two threads.